File: web-spring/src/main/java/io/github/agussyahrilmubarok/web/application/dto/user/UpdateUserRequest.java
package io.github.agussyahrilmubarok.web.application.dto.user;

import jakarta.validation.constraints.Email;
import jakarta.validation.constraints.Size;
import java.util.UUID;

public record UpdateUserRequest(
        UUID id,

        @Size(max = 100, message = "Name must be between 2 and 100 characters")
        String name,

        @Email(message = "Email format is invalid") @Size(max = 150, message = "Email must not exceed 150 characters")
        String email,

        @Size(min = 8, max = 72, message = "Password must be between 8 and 72 characters")
        String password) {}

File: examples/web-mvc-i18n/web-dotnet/Web/Application/DTO/User/UpdateUserRequest.cs
using System.ComponentModel.DataAnnotations;

namespace Web.Application.DTO.User;

public class UpdateUserRequest
{
    [StringLength(100, MinimumLength = 2, ErrorMessage = "Name must be between 2 and 100 characters")]
    public string? Name { get; set; }

    [ValidEmailIfPresent]
    [StringLength(150)]
    public string? Email { get; set; }

    [StringLength(72, MinimumLength = 8, ErrorMessage = "Password must be between 8 and 72 characters")]
    public string? Password { get; set; }

    public UpdateUserRequest()
    {
    }

    public UpdateUserRequest(string name, string email)
    {
        Name = name;
        Email = email;
    }
}

public class ValidEmailIfPresentAttribute : ValidationAttribute
{
    protected override ValidationResult? IsValid(object? value, ValidationContext validationContext)
    {
        if (value is null || string.IsNullOrWhiteSpace(value.ToString()))
            return ValidationResult.Success;

        var isValid = new EmailAddressAttribute().IsValid(value.ToString());
        return isValid
            ? ValidationResult.Success
            : new ValidationResult("Email is not valid");
    }
}